cli/src/cli_util.rs

```diff
@@ -1157,8 +1159,52 @@ impl WorkspaceCommandHelper {
         self.check_repo_rewritable(self.repo().as_ref(), commits)
     }
 
+    fn reset_author_revset(&self) -> Result<Rc<RevsetExpression>, CommandError> {
+        self.reset_author_revset
+            .get_or_try_init(|| {
+                let reset_author = self
+                    .settings
+                    .config()
+                    .get_string("revsets.reset-author-on-edit")
+                    .unwrap_or_default();
+                revset::parse(&reset_author, &self.revset_parse_context()).map_err(|e| {
+                    config_error_with_message("Invalid `revsets.reset-author-on-edit`", e)
+                })
+            })
+            .cloned()
+    }
+
+    fn evaluate_reset_author_revset(
+        mut expression: RevsetExpressionEvaluator,
+        commit_id: &CommitId,
+    ) -> Result<bool, CommandError> {
+        expression.intersect_with(&RevsetExpression::commit(commit_id.clone()));
+        Ok(expression
+            .evaluate_to_commit_ids()
+            .map_err(|e| config_error_with_message("Invalid `revsets.reset-author-on-edit`", e))?
+            .next()
+            .is_some())
+    }
+
+    pub fn check_reset_author(
+        &self,
+        repo: &dyn Repo,
+        commit_id: &CommitId,
+    ) -> Result<bool, CommandError> {
+        let expression = RevsetExpressionEvaluator::new(
+            repo,
+            self.revset_extensions.clone(),
+            self.id_prefix_context()?,
+            self.reset_author_revset()?,
+        );
+        Self::evaluate_reset_author_revset(expression, commit_id)
+    }
+
     #[instrument(skip_all)]
     fn snapshot_working_copy(&mut self, ui: &mut Ui) -> Result<(), CommandError> {
+        let reset_author_revset = self.reset_author_revset()?;
+        let id_prefix_context = self.id_prefix_context()?.clone();
```

Comment:
Cloning `IdPrefixContext` can be expensive. Perhaps, some part of `self.id_prefix_context()` can be moved to `ReadonlyUserRepo` to get around borrow checker.

BTW, `self.user_repo` is reassigned in this function, so obtaining `id_prefix_context` here is wrong.